app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java

```diff
@@ -95,6 +138,100 @@ public void afterTextChanged(final Editable s) {
+        super.onFinishInflate();
+        _defaultPaddingLeft = getPaddingLeft();
+        _paint.setTextAlign(Paint.Align.RIGHT);
+    }
+
+    @Override
+    protected void onAttachedToWindow() {
+        super.onAttachedToWindow();
+        _scrollView = getParent() instanceof ScrollView ? (ScrollView) getParent() : (ScrollView) getParent().getParent();
+        _scrollView.setOnScrollChangeListener((v, scrollX, scrollY, oldScrollX, oldScrollY) -> {
+            if (_firstVisibleLineUpdated) {
+                _firstVisibleLineUpdated = false;
+            }
+        });
+    }
+
+    @Override
+    public boolean onPreDraw() {
+        _paint.setTextSize(getTextSize());
+        return super.onPreDraw();
+    }
+
+    @Override
+    protected void onDraw(Canvas canvas) {
+        super.onDraw(canvas);
+        // If line numbers can be drawn
+        if (_nuEnabled && _maxLineNumber < (AppSettings._isDeviceGoodHardware ? 5000 : 3000)) {
+            drawLineNumbers(canvas);
+        } else if (getPaddingLeft() != _defaultPaddingLeft) {
+            _maxLineNumberWidth = 0;
+            // Reset padding without line numbers fence
+            setPadding(_defaultPaddingLeft, getPaddingTop(), getPaddingRight(), getPaddingBottom());
+        }
+    }
+
+    private void drawLineNumbers(Canvas canvas) {
+        final Editable text = getText();
+        final Layout layout = getLayout();
+        final float offsetY = getPaddingTop();
+        final int top = _scrollView.getScrollY() - 100; // Top of current visible area
+        final int bottom = _scrollView.getScrollY() + _scrollView.getHeight(); // Bottom of current visible area
+        final int width = (int) _paint.measureText(String.valueOf(_maxLineNumber));
+        if (_maxLineNumberWidth != width) {
+            _maxLineNumberWidth = width;
+            _x = LINE_NUMBERS_PADDING_LEFT + width;
+            setPadding(_x + LINE_NUMBERS_PADDING_RIGHT + 10, getPaddingTop(), getPaddingRight(), getPaddingBottom());
+        }
+
+        // Draw the right border
+        _paint.setColor(Color.LTGRAY);
+        canvas.drawLine(_x + LINE_NUMBERS_PADDING_RIGHT, top, _x + LINE_NUMBERS_PADDING_RIGHT, bottom, _paint);
+
+        // Draw first line number
+        _paint.setColor(Color.GRAY);
+        canvas.drawText(String.valueOf(1), _x, layout.getLineBounds(0, null) + offsetY, _paint);
+
+        // Draw other line numbers
+        if (text != null) {
+            final int count = getLineCount();
+            int i = 1, number = 1;
+
+            if (_firstVisibleLineUpdated) {
+                // Set and then iterate from the first visible line
+                i = _firstVisibleLine[0];
+                number = _firstVisibleLine[1];
+            } else {
+                // Set the first visible line invalid, it needs to be updated
+                _firstVisibleLine[0] = -1;
+                _firstVisibleLine[1] = -1;
+            }
+
+            for (int y; i < count; i++) {
```

Comment:
since y is not the loop variable, I think it should not be declared here.

Just do 

`for (; i < count; i++)`

`final int y = layout.getLineBounds(i, null);`

Reply:
I am attempting to make these changes in a local branch. If they work I will push it so you can pull them in if you think they work